Create CalculatorTest Module

chicago4550 opened the pull request:
# Changelist
- Created new module that allows the bot to take in a simple operation, only including addition, subtraction, multiplication and division. The module only allows for one calculation per command and if the format is not followed, (number operation number) the bot will say so and ask you to reenter your command using the correct format. This simple calculator can deal with both decimals and negative numbers, as all numbers are stored as floats

## Type of change
- [ ] New feature (non-breaking change which adds functionality)


## Local configuration:
- OS: (Windows)
- JDK: (OpenJDK on Java 8)

chicago4550 committed:
Create CalculatiorTest Module

This module allows the bod to take in a simple operation, only including addition, subtraction, multiplication and division. The module only allows for one calculation per command and if the format is not followed, (number operation number) the bot will say so and ask you to reenter your command using the correct format. This simple calculator can deal with both decimals and negative numbers, as all numbers are stored as floats

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+
+
+@Service
+public final class ModuleCalculatorTest extends Module {
```
This class should be renamed to ModuleCalculator and moved out of the `ghost2.test` module. This is a command, not a unit test.

chicago4550 replied:
Thank you for the quick reply. Now that I am running it again today in IntelliJ instead of atom and PowerShell, I am getting errors on MacAdapter.class not being able to be in a top-level directory. Where is MacAdapter.class, or is that an issue with how I have everything configured on my end? In addition, runtimeBuilder cannot find my OpenJDK Java.exe, despite it pointing directly to it. Any ideas on how to fix these two issues?

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+public final class ModuleCalculatorTest extends Module {
+
+  @ReflectiveAccess
+  public ModuleUrbanDictionary(){
```
Doesn't compile: the constructor should use the name of the class.
https://docs.oracle.com/javase/tutorial/java/javaOO/constructors.html

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+  }
+  @Override
+  @ReflectiveAccess
+  public void involk(@NotNull final CommandContext ctx) {
```
Doesn't compile: `involk` is not a method in `Module`. Implement `invoke`.
Also, you need to import `NotNull`.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+  @ReflectiveAccess
+  public void involk(@NotNull final CommandContext ctx) {
+
+        String[] testNewStrings = ctx;
```
Doesn't compile: `ctx` is a reference to a `CommandContext` object. You can't assign it to a `String[]`. Try `CommandContext.getArgs` and use a `List<String>`.
This needs a better name than `testNewStrings`.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+  public ModuleUrbanDictionary(){
+    super(new ModuleInfo.Builder(ModuleCalculatorTest.class)
+
+        .WithName("calculator")
```
Doesn't compile: `WithName` is not a method of `ModuleInfo.Builder`. Method names are case-sensitive.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+            ctx.getChannel().createMessage("this calculation is not in the correct format, please follow the format of (number calculation number)")
+        }
+
+        Boolean testBoolean = false;
```
This doesn't need to be a `Boolean`. Use the primitive.
Also needs a more descriptive variable name.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+
+        Boolean testBoolean = false;
+        for (int i = 0; i < testNewStrings.length; i++) {
+            if (i % 2 == 0 && testBoolean == false) {
```
`testBoolean == false` can be simplified to `!testBoolean`

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+                    testBoolean = true;
+                }
+            }
+            else if (testBoolean == false && i != testNewStrings.length - 1) {
```
`testBoolean == false` can be simplified to `!testBoolean`

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+            if (i % 2 == 0 && testBoolean == false) {
+                try {
+                    float f = Float.parseFloat(testNewStrings[i]);
+                    testStrings.add(String.valueOf(f));
```
Doesn't compile: `testStrings` doesn't exist in this scope, and arrays do not have an `add` method.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+        .WithName("calculator")
+        .withAliases("calc")
+        .withDescription("Complete a 2-number calculation with either addition, subtraction, multiplication, or division")
+        .withBotPermissions());
```
Doesn't compile: `withBotPermissions` expects a `PermissionSet`

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+        float num2 = Float.parseFloat((String) testStrings.get(2));
+
+        if(testStrings.get(1).compareTo("+") == 0){
+            result = ad(num1, num2);
```
Doesn't compile: `ad` is a non-existant method. This can be replaced with an operator.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+            result = ad(num1, num2);
+        }
+        else if(testStrings.get(1).compareTo("-") == 0){
+            result = sub(num1, num2);
```
Doesn't compile: `sub` is a non-existant method. This can be replaced with an operator.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+            result = sub(num1, num2);
+        }
+        else if(testStrings.get(1).compareTo("*") == 0){
+            result = mult(num1, num2);
```
Doesn't compile: `mult` is a non-existant method. This can be replaced with an operator.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+            result = mult(num1, num2);
+        }
+        else if(testStrings.get(1).compareTo("/") == 0){
+            result = div(num1, num2);
```
Doesn't compile: `div` is a non-existant method. This can be replaced with an operator.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+        float num1 = Float.parseFloat((String) testStrings.get(0));
+        float num2 = Float.parseFloat((String) testStrings.get(2));
+
+        if(testStrings.get(1).compareTo("+") == 0){
```
Rather than using a string of if/elseifs here, you can switch on `testStrings.get(1)`. Much cleaner.

zocat64 commented on src/test/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculatorTest.java:
```diff
@@ -0,0 +1,83 @@
+        for (int i = 0; i < testNewStrings.length; i++) {
+            if (i % 2 == 0 && testBoolean == false) {
+                try {
+                    float f = Float.parseFloat(testNewStrings[i]);
```
Use a double instead of a float, memory is not that much of a concern and this gets GCed when `invoke` finishes anyways

zocat64 reviewed (changes requested):
Thanks for the PR! There's a lot of pretty glaring issues here. I can tell you're relatively new to Java and/or programming in general, so here's my thoughts for you:
- Work on your own projects for a while until you're comfortable contributing to more complex open-source projects. Don't rush it.
- Download an IDE -- IntelliJ is free! Half of these errors wouldn't be here if you had one.
- Make sure your code compiles before you open a PR!

For now, fix the issues I've mentioned in the review and I'll work on cleaning this up and merging.

chicago4550 committed:
Renamed variables and fixed compiling issues

I moved the file out of the test area and into the main part of the code, as well as changing variable names to make them relevent to the code and fixing all of the other compiling issues

chicago4550 committed:
Update ModuleCalculator to remove commented out and unused code

I just deleted code that was not being used and was just commened out

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+
+
+
+@Service
```
This module shouldn't be a Spring service.

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+        double num1 = Double.parseDouble((String) Args.get(0));
+        double num2 = Double.parseDouble((String) Args.get(2));
+
+        if(Args.get(1).compareTo("+") == 0){
+            result = num1 + num2;
+        }
+        else if(Args.get(1).compareTo("-") == 0){
+            result = num1 - num2;
+        }
+        else if(Args.get(1).compareTo("*") == 0){
+            result = num1 * num2;
+        }
+        else if(Args.get(1).compareTo("/") == 0){
+            result = num1 / num2;
+        }
```
Convert these to a switch statement.

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+
+  }
+
+    /**
+     * Constructs a new Module. Should only be used by subclasses of Module.
+     *
+     * @param info ModuleInfo.Builder with the subclass' metadata
+     */
+    protected ModuleCalculator(ModuleInfo.Builder info) {
+        super(info);
+    }
```
Don't implement a protected constructor here. This class is final anyways.

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+            ctx.getChannel().createMessage("this calculation is not in the correct format, please follow the format of (number calculation number)");
+        }
+
+        boolean incorrectInput = false;
+        for (int i = 0; i < Args.size(); i++) {
+            if (i % 2 == 0 && !incorrectInput) {
+                try {
+                    double f = Double.parseDouble(Args.get(i));
+                } catch (Exception e) {
+                    ctx.getChannel().createMessage("You entered an invalid number, please try this command again with a valid number");
+                    incorrectInput = true;
+                }
+            }
+
+        }
```
Using a loop here is overengineering it and is less readable -- just fetch the operands by index and try to parse them.

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+        else if(Args.get(1).compareTo("/") == 0){
+            result = num1 / num2;
+        }
+        ctx.getChannel().createMessage("result is " + result);
```
Replace with `ctx.replyBlocking(...)`

zocat64 commented on src/main/java/com/github/coleb1911/ghost2/commands/modules/fun/ModuleCalculator.java:
```diff
@@ -0,0 +1,84 @@
+  @ReflectiveAccess
+  public ModuleCalculator(){
+    super(new ModuleInfo.Builder(ModuleCalculator.class)
+            .withName("Calculator")
```
Command names should be lowercase.

zocat64 reviewed (changes requested):
There's a few more things of note left.

cbryant02 commented:
On the MacAdapter error: that's not a class that exists in this project, so I'm not sure.
I'm not sure what "runtime builder" you're referring to, but if your JDK is unconfigured, there should be a message at the top of the IDE that lets you set it up.

cbryant02 commented:
Going to close this due to inactivity. Feel free to reopen it any time.
